### ISSUE TITLE
EMQ-X Bridge only sends first keepalive packet causing disconnects

### ISSUE THREAD

**Issue by NielsH** [emqx/emqx#3404]
Hi,

I have configured two EMQ-X 4.0.5 Clusters and would like to put a bridge between. I'm fairly new to EMQ-X so perhaps it is a configuration issue, but we're having issues which appear to be related to keepalive_timeout.

Environment:
```
emqx_ctl status
Node 'emqx@172.31.164.18' is started
emqx 4.0.5 is running

lscpu
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
Address sizes:       46 bits physical, 48 bits virtual
CPU(s):              2
On-line CPU(s) list: 0,1
Thread(s) per core:  1
Core(s) per socket:  1
Socket(s):           2
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Gold 5218 CPU @ 2.30GHz
Stepping:            7
CPU MHz:             2300.249
BogoMIPS:            4600.06
Hypervisor vendor:   Xen
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            22528K
NUMA node0 CPU(s):   0,1
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush acpi mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl cpuid pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single pti intel_ppin ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx rdseed adx smap clflushopt clwb xsaveopt xsavec xgetbv1 xsaves pku ospke md_clear flush_l1d


cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

Kernel: 4.19.0-8-amd64 #1 SMP Debian 4.19.98-1 (2020-01-26) x86_64 GNU/Linux


```

Little overview:
Cluster A:
- 172.31.164.18
- 172.31.164.19
- 172.31.164.20

Cluster B:
- 172.31.165.18
- 172.31.165.19
- 172.31.165.20

Eventually we want to bridge like this:
[All nodes in cluster A] --> Loadbalancer --> [Random node in Cluster B]
[All nodes in cluster B] --> Loadbalancer --> [Random node in Cluster A]

However, for debugging purposes I have now configured it like this:
Cluster B 172.31.165.20 --> Cluster A 172.31.164.18

The issue we are having is that the bridge goes up, and connects correctly, however it appears to only send a single keepalive packet. After that, it stops doing that, and when 2 keepalives are missed the bridge is disconnected. I have experimented with higher/lower settings, and the lower the keepalive, the earlier it disconnects (logical of course because it will be faster to miss 2 keepalive packets)

This is the bridge config:
```
bridge.mqtt.tst.address = 172.31.164.18:1883
bridge.mqtt.tst.username = bridge_tst_
bridge.mqtt.tst.password = <removed>
bridge.mqtt.tst.forwards = log/#

bridge.mqtt.tst.proto_ver = mqttv5
bridge.mqtt.tst.forward_mountpoint = bridge/tst/${node}/
bridge.mqtt.tst.queue.replayq_dir = /var/lib/emqx/emqx_tst_bridge/
bridge.mqtt.tst.queue.replayq_seg_bytes = 10MB
bridge.mqtt.tst.queue.max_total_size = 5GB

bridge.mqtt.tst.start_type = manual
bridge.mqtt.tst.bridge_mode = true
bridge.mqtt.tst.clean_start = true
bridge.mqtt.tst.keepalive = 20s
bridge.mqtt.tst.reconnect_interval = 10s
bridge.mqtt.tst.retry_interval = 5s
bridge.mqtt.tst.batch_size = 32
bridge.mqtt.tst.max_inflight_size = 32
```

Here is a tcpdump from Cluster B 172.31.165.20 (the node with the bridge config):
```
2020-04-21 15:00:58.625422 IP 172.31.165.20.45207 > 172.31.164.18.1883: Flags [S], seq 1262955451, win 64240, options [mss 1460,sackOK,TS val 1613060145 ecr 0,nop,wscale 7], length 0
2020-04-21 15:00:58.625903 IP 172.31.164.18.1883 > 172.31.165.20.45207: Flags [S.], seq 62892336, ack 1262955452, win 65160, options [mss 1460,sackOK,TS val 345021343 ecr 1613060145,nop,wscale 7], length 0
2020-04-21 15:00:58.625919 IP 172.31.165.20.45207 > 172.31.164.18.1883: Flags [.], ack 1, win 502, options [nop,nop,TS val 1613060145 ecr 345021343], length 0
2020-04-21 15:00:58.626148 IP 172.31.165.20.45207 > 172.31.164.18.1883: Flags [P.], seq 1:52, ack 1, win 502, options [nop,nop,TS val 1613060145 ecr 345021343], length 51
2020-04-21 15:00:58.626408 IP 172.31.164.18.1883 > 172.31.165.20.45207: Flags [.], ack 52, win 509, options [nop,nop,TS val 345021344 ecr 1613060145], length 0
2020-04-21 15:00:58.629198 IP 172.31.164.18.1883 > 172.31.165.20.45207: Flags [P.], seq 1:72, ack 52, win 509, options [nop,nop,TS val 345021346 ecr 1613060145], length 71
2020-04-21 15:00:58.629215 IP 172.31.165.20.45207 > 172.31.164.18.1883: Flags [.], ack 72, win 502, options [nop,nop,TS val 1613060148 ecr 345021346], length 0
2020-04-21 15:01:18.630510 IP 172.31.165.20.45207 > 172.31.164.18.1883: Flags [P.], seq 52:54, ack 72, win 502, options [nop,nop,TS val 1613080149 ecr 345021346], length 2
2020-04-21 15:01:18.631724 IP 172.31.164.18.1883 > 172.31.165.20.45207: Flags [P.], seq 72:74, ack 54, win 509, options [nop,nop,TS val 345041349 ecr 1613080149], length 2
2020-04-21 15:01:18.631743 IP 172.31.165.20.45207 > 172.31.164.18.1883: Flags [.], ack 74, win 502, options [nop,nop,TS val 1613080151 ecr 345041349], length 0
2020-04-21 15:01:58.632495 IP 172.31.165.20.45207 > 172.31.164.18.1883: Flags [P.], seq 54:56, ack 74, win 502, options [nop,nop,TS val 1613120151 ecr 345041349], length 2
2020-04-21 15:01:58.633233 IP 172.31.164.18.1883 > 172.31.165.20.45207: Flags [P.], seq 74:78, ack 56, win 509, options [nop,nop,TS val 345081349 ecr 1613120151], length 4
2020-04-21 15:01:58.633278 IP 172.31.165.20.45207 > 172.31.164.18.1883: Flags [.], ack 78, win 502, options [nop,nop,TS val 1613120151 ecr 345081349], length 0
2020-04-21 15:01:58.633399 IP 172.31.164.18.1883 > 172.31.165.20.45207: Flags [F.], seq 78, ack 56, win 509, options [nop,nop,TS val 345081349 ecr 1613120151], length 0
2020-04-21 15:01:58.633491 IP 172.31.165.20.45207 > 172.31.164.18.1883: Flags [F.], seq 56, ack 79, win 502, options [nop,nop,TS val 1613120152 ecr 345081349], length 0
2020-04-21 15:01:58.633662 IP 172.31.164.18.1883 > 172.31.165.20.45207: Flags [.], ack 57, win 509, options [nop,nop,TS val 345081350 ecr 1613120152], length 0

16 packets captured
16 packets received by filter
0 packets dropped by kerne
```

And the logfile at Cluster B 172.31.165.20:
```
2020-04-21 15:00:58.625 [debug] emqtt(): SEND Data: {mqtt_packet,{mqtt_packet_header,1,false,0,false},{mqtt_packet_connect,<<"MQTT">>,5,true,true,false,0,false,20,#{},<<>>,undefined,undefined,undefined,<<"bridge_tst_">>,<<"<removed password>">>},undefined}
2020-04-21 15:00:58.629 [debug] emqtt(): RECV Data: <<32,69,0,0,66,18,0,47,77,106,107,121,79,68,77,51,77,106,
                      99,50,79,68,77,120,78,68,65,122,79,84,65,53,77,106,89,
                      120,78,106,77,51,79,68,107,120,77,84,77,119,78,106,85,48,
                      78,122,67,39,0,16,0,0,37,1,42,1,41,1,34,255,255,40,1>>
2020-04-21 15:00:58.629 [info] [Bridge] Bridge emqx_bridge_worker_tst is connecting......
2020-04-21 15:01:18.630 [debug] emqtt(MjkyODM3Mjc2ODMxNDAzOTA5MjYxNjM3ODkxMTMwNjU0NzC): SEND Data: {mqtt_packet,{mqtt_packet_header,12,false,0,false},undefined,undefined}
2020-04-21 15:01:18.631 [debug] emqtt(MjkyODM3Mjc2ODMxNDAzOTA5MjYxNjM3ODkxMTMwNjU0NzC): RECV Data: <<208,0>>
2020-04-21 15:01:58.632 [debug] emqtt(MjkyODM3Mjc2ODMxNDAzOTA5MjYxNjM3ODkxMTMwNjU0NzC): SEND Data: {mqtt_packet,{mqtt_packet_header,12,false,0,false},undefined,undefined}
2020-04-21 15:01:58.633 [debug] emqtt(MjkyODM3Mjc2ODMxNDAzOTA5MjYxNjM3ODkxMTMwNjU0NzC): RECV Data: <<224,2,141,
                                                                     0>>
2020-04-21 15:01:58.633 [info] [Bridge] Bridge emqx_bridge_worker_tst diconnected
reason={141,#{}}
2020-04-21 15:01:58.633 [error] ** State machine <0.2871.0> terminating
** Last event = {cast,{mqtt_packet,{mqtt_packet_header,14,false,0,false},
                                   {mqtt_packet_disconnect,141,#{}},
                                   undefined}}
** When server state  = {connected,
                            {state,undefined,<0.1599.0>,
                                #{disconnected =>
                                      #Fun<emqx_bridge_mqtt.3.66057189>,
                                  puback => #Fun<emqx_bridge_mqtt.1.66057189>,
                                  publish =>
                                      #Fun<emqx_bridge_mqtt.2.66057189>},
                                "172.31.164.18",1883,[],emqtt_sock,
                                #Port<0.344>,[],60000,true,
                                <<"MjkyODM3Mjc2ODMxNDAzOTA5MjYxNjM3ODkxMTMwNjU0NzC">>,
                                true,<<"bridge_tst_">>,
                                <<"<removed password>">>,5,<<"MQTT">>,20,
                                #Ref<0.2053434580.3850108930.93972>,false,
                                false,false,
                                {mqtt_msg,0,false,false,undefined,undefined,
                                    undefined,undefined},
                                #{'Assigned-Client-Identifier' =>
                                      <<"MjkyODM3Mjc2ODMxNDAzOTA5MjYxNjM3ODkxMTMwNjU0NzC">>,
                                  'Maximum-Packet-Size' => 1048576,
                                  'Retain-Available' => 1,
                                  'Shared-Subscription-Available' => 1,
                                  'Subscription-Identifier-Available' => 1,
                                  'Topic-Alias-Maximum' => 65535,
                                  'Wildcard-Subscription-Available' => 1},
                                [],#{},infinity,#{},#{},true,30000,undefined,
                                5000000,undefined,0,4,
                                {none,
                                    #{max_size => 268435455,
                                      strict_mode => false,version => 5}}}}
** Reason for termination = exit:{disconnected,141,#{}}
** Callback mode = state_functions
** Stacktrace =
**  [{gen_statem,loop_state_callback_result,11,
                 [{file,"gen_statem.erl"},{line,1298}]},
     {proc_lib,wake_up,3,[{file,"proc_lib.erl"},{line,259}]}]

2020-04-21 15:01:58.634 [error]   crasher:
    initial call: emqtt:init/1
    pid: <0.2871.0>
    registered_name: []
    exception exit: {disconnected,141,#{}}
      in function  gen_statem:loop_state_callback_result/11 (gen_statem.erl, line 1298)
    ancestors: [emqx_bridge_worker_tst,emqx_bridge_mqtt_sup,<0.1597.0>]
    message_queue_len: 0
    messages: []
    links: [<0.1599.0>]
    dictionary: [{send_oct,53}]
    trap_exit: true
    status: running
    heap_size: 6772
    stack_size: 27
    reductions: 46004
  neighbours:

```

On the receiving side, where the bridge is pointing to, we see this in the logfiles:
```
2020-04-21 15:00:58.626 [debug] 172.31.165.20:45207 [MQTT] RECV <<16,49,0,4,77,81,84,84,133,194,0,20,0,0,0,0,12,98,114,105,100,103,
              101,95,100,114,105,115,95,0,20,56,56,105,67,98,77,72,116,82,101,
              54,51,78,81,122,69,107,52,106,80>>
2020-04-21 15:00:58.626 [debug] 172.31.165.20:45207 [MQTT] RECV CONNECT(Q0, R0, D0, ClientId=, ProtoName=MQTT, ProtoVsn=5, CleanStart=true, KeepAlive=20, Username=bridge_tst_, Password=******)
2020-04-21 15:00:58.628 [debug] <<"MjkyODM3Mjc2ODMxNDAzOTA5MjYxNjM3ODkxMTMwNjU0NzC">>@172.31.165.20:45207 [MQTT] SEND CONNACK(Q0, R0, D0, AckFlags=0, ReasonCode=0)
2020-04-21 15:01:18.630 [debug] <<"MjkyODM3Mjc2ODMxNDAzOTA5MjYxNjM3ODkxMTMwNjU0NzC">>@172.31.165.20:45207 [MQTT] RECV <<192,0>>
2020-04-21 15:01:18.631 [debug] <<"MjkyODM3Mjc2ODMxNDAzOTA5MjYxNjM3ODkxMTMwNjU0NzC">>@172.31.165.20:45207 [MQTT] SEND PINGRESP(Q0, R0, D0)

2020-04-21 15:01:58.632 [debug] <<"MjkyODM3Mjc2ODMxNDAzOTA5MjYxNjM3ODkxMTMwNjU0NzC">>@172.31.165.20:45207 [MQTT] SEND DISCONNECT(Q0, R0, D0, ReasonCode=141)
2020-04-21 15:01:58.633 [debug] <<"MjkyODM3Mjc2ODMxNDAzOTA5MjYxNjM3ODkxMTMwNjU0NzC">>@172.31.165.20:45207 [MQTT] Force to close the socket due to keepalive_timeout
2020-04-21 15:01:58.633 [debug] <<"MjkyODM3Mjc2ODMxNDAzOTA5MjYxNjM3ODkxMTMwNjU0NzC">>@172.31.165.20:45207 [MQTT] Terminated due to {shutdown,keepalive_timeout}
```

Timeline wise you can see:
15:00:58 - Bridge is started. Shown in tcpdump and EMQ-X logs on both servers.
15:01:18 - First keepalive is sent from `172.31.165.20.45207 > 172.31.164.18.1883` - as you can see in the tcpdump and logs on both side. So keepalive packet is sent, and received properly.
15:01:38 - First keepalive packet missed. Not present in logs nor tcpdump
15:01:58 - Second keepalive packet missed. Not present in logs nor tcpdump
15:01:58 - Connection is closed by 172.31.164.18 as shown in logs on both sides. The receiving side gives the reason: keepalive_timeout.

Are we doing something wrong here with the settings, or is this a bug?

Cheers,
Niels

**Comment by turtleDeng** [comment]
Thanks for your feedback, we will check

**Comment by zhanghongtong** [comment]
Hi, @NielsH 
Thanks for your feedback, we will try to reproduce this issue and fix it.

**Comment by qingchuwudi** [comment]
This bug has been fix in [v4.0.6](https://github.com/emqx/emqx/releases/tag/v4.0.6)
